src/librustc_lint/types.rs

```diff
@@ -957,14 +963,20 @@ impl<'a, 'tcx> ImproperCTypesVisitor<'a, 'tcx> {
         // it is only OK to use this function because extern fns cannot have
         // any generic types right now:
         let ty = self.cx.tcx.normalize_erasing_regions(ParamEnv::reveal_all(), ty);
-        // C doesn't really support passing arrays by value.
-        // The only way to pass an array by value is through a struct.
-        // So we first test that the top level isn't an array,
-        // and then recursively check the types inside.
+
+        // C doesn't really support passing arrays by value - the only way to pass an array by value
+        // is through a struct. So, first test that the top level isn't an array, and then
+        // recursively check the types inside.
         if !is_static && self.check_for_array_ty(sp, ty) {
             return;
         }
 
+        // Don't report FFI errors for unit return types - not checking in `check_foreign_fn` so
```

Comment:
I can't quite parse this sentence. Is it "we are now checking in"?

Reply:
Apparently this made sense to me when I wrote it - I understand past-me to have meant "we aren't checking in `check_foreign_fn` (where this check might have been expected) because we want to check after normalization".

Reply:
Pushed a hopefully clearer comment.